### PR TITLE
[Fix] add type hints of part 8 including iou_loss.py, etc.

### PR DESCRIPTION
**Pull request by lihua199710**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Finish the task of Codecamp to add type hints of part 8 including the following files:
1. models/losses/iou_loss.py; 
2. models/losses/kd_loss.py
3. models/losses/mse_loss.py
4. models/losses/seesaw_loss.py
5. models/losses/smooth_l1_loss.py
6. models/losses/utils.py
7. models/losses/varifocal_loss.py

## Modification

I have only modified the following files:
1. models/losses/iou_loss.py; 
2. models/losses/kd_loss.py
3. models/losses/mse_loss.py
4. models/losses/seesaw_loss.py
5. models/losses/smooth_l1_loss.py
6. models/losses/utils.py
7. models/losses/varifocal_loss.py

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [x] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
- [x] The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet/models/losses/iou_loss.py [hunk 1]
```diff
@@ -59,8 +67,11 @@ def bounded_iou_loss(pred, target, beta=0.2, eps=1e-3):
     Args:
         pred (torch.Tensor): Predicted bboxes.
         target (torch.Tensor): Target bboxes.
-        beta (float): beta parameter in smoothl1.
-        eps (float): eps to avoid NaN.
+        beta (float, optional): beta parameter in smoothl1.
+        eps (float, optional): eps to avoid NaN.
```
Comment (ZwwWayne) [inline mmdet/models/losses/iou_loss.py:71]: Epsilon to avoid Nan values

File: mmdet/models/losses/iou_loss.py [hunk 2]
```diff
@@ -11,7 +12,11 @@
 
 
 @weighted_loss
-def iou_loss(pred, target, linear=False, mode='log', eps=1e-6):
+def iou_loss(pred: torch.Tensor,
```
Comment (BIGWangYuDong) [inline mmdet/models/losses/iou_loss.py:15]: @ZwwWayne is that necessary to use `Tensor` here? we can use `from torch import Tensor` at beginning

Comment (ZwwWayne) [reply mmdet/models/losses/iou_loss.py:15]: We can do that.

File: mmdet/models/losses/iou_loss.py (hunk 1 above)
Comment (lihua199710) [reply mmdet/models/losses/iou_loss.py]: I have modified it.

File: mmdet/models/losses/iou_loss.py [hunk 3]
```diff
@@ -95,15 +107,15 @@ def bounded_iou_loss(pred, target, beta=0.2, eps=1e-3):
 
 
 @weighted_loss
-def giou_loss(pred, target, eps=1e-7):
+def giou_loss(pred: Tensor, target: Tensor, eps: float = 1e-7) -> Tensor:
     r"""`Generalized Intersection over Union: A Metric and A Loss for Bounding
     Box Regression <https://arxiv.org/abs/1902.09630>`_.
 
     Args:
         pred (torch.Tensor): Predicted bboxes of format (x1, y1, x2, y2),
```
Comment (BIGWangYuDong) [inline mmdet/models/losses/iou_loss.py:115]: torch.Tensor -> Tensor

Comment (lihua199710) [reply mmdet/models/losses/iou_loss.py]: Thanks, i have corrected it.

File: mmdet/models/losses/iou_loss.py [hunk 4]
```diff
@@ -320,18 +335,18 @@ def __init__(self,
...
-                target,
-                weight=None,
-                avg_factor=None,
-                reduction_override=None,
+                pred: Tensor,
+                target: Tensor,
+                weight: Optional[Tensor] = None,
+                avg_factor: Optional[int] = None,
+                reduction_override: Optional[str] = None,
                 **kwargs):
```
Comment (BIGWangYuDong) [inline mmdet/models/losses/iou_loss.py:343]:
```suggestion
                **kwargs) -> Tensor:
```

Not sure, but need have a check. Also, Docstrings need update. Such as 
```
Returns: 
    Tensor: xxx
```

File: mmdet/models/losses/iou_loss.py [hunk 5]
```diff
@@ -293,20 +308,20 @@ class IoULoss(nn.Module):
...
-                 eps=1e-6,
-                 reduction='mean',
-                 loss_weight=1.0,
-                 mode='log'):
-        super(IoULoss, self).__init__()
+                 linear: bool = False,
+                 eps: float = 1e-6,
+                 reduction: str = 'mean',
+                 loss_weight: float = 1.0,
+                 mode: str = 'log'):
```
Comment (BIGWangYuDong) [inline mmdet/models/losses/iou_loss.py:323]:
```suggestion
                 mode: str = 'log') -> None:
```

File: mmdet/models/losses/iou_loss.py [hunk 6]
```diff
@@ -368,19 +383,23 @@ def forward(self,
 @MODELS.register_module()
 class BoundedIoULoss(nn.Module):
 
-    def __init__(self, beta=0.2, eps=1e-3, reduction='mean', loss_weight=1.0):
-        super(BoundedIoULoss, self).__init__()
+    def __init__(self,
+                 beta: float = 0.2,
+                 eps: float = 1e-3,
+                 reduction: str = 'mean',
+                 loss_weight: float = 1.0):
```
Comment (BIGWangYuDong) [inline mmdet/models/losses/iou_loss.py:390]: -> None

File: mmdet/models/losses/iou_loss.py [hunk 7]
```diff
@@ -368,19 +383,23 @@ def forward(self,
...
-                target,
-                weight=None,
-                avg_factor=None,
-                reduction_override=None,
+                pred: Tensor,
+                target: Tensor,
+                weight: Optional[Tensor] = None,
+                avg_factor: Optional[int] = None,
+                reduction_override: Optional[str] = None,
                 **kwargs):
```
Comment (BIGWangYuDong) [inline mmdet/models/losses/iou_loss.py:403]: -> Tensor

File: mmdet/models/losses/iou_loss.py [hunk 8]
```diff
@@ -404,18 +423,21 @@ def forward(self,
 @MODELS.register_module()
 class GIoULoss(nn.Module):
 
-    def __init__(self, eps=1e-6, reduction='mean', loss_weight=1.0):
-        super(GIoULoss, self).__init__()
+    def __init__(self,
+                 eps: float = 1e-6,
+                 reduction: str = 'mean',
+                 loss_weight: float = 1.0):
```
Comment (BIGWangYuDong) [inline mmdet/models/losses/iou_loss.py:429]: -> None

File: mmdet/models/losses/iou_loss.py [hunk 9]
```diff
@@ -404,18 +423,21 @@ def forward(self,
...
-                target,
-                weight=None,
-                avg_factor=None,
-                reduction_override=None,
+                pred: Tensor,
+                target: Tensor,
+                weight: Optional[Tensor] = None,
+                avg_factor: Optional[int] = None,
+                reduction_override: Optional[str] = None,
                 **kwargs):
```
Comment (BIGWangYuDong) [inline mmdet/models/losses/iou_loss.py:441]: -> Tensor

File: mmdet/models/losses/iou_loss.py [hunk 10]
```diff
@@ -444,18 +466,21 @@ def forward(self,
 @MODELS.register_module()
 class DIoULoss(nn.Module):
 
-    def __init__(self, eps=1e-6, reduction='mean', loss_weight=1.0):
-        super(DIoULoss, self).__init__()
+    def __init__(self,
+                 eps: float = 1e-6,
+                 reduction: str = 'mean',
+                 loss_weight: float = 1.0):
```
Comment (BIGWangYuDong) [inline mmdet/models/losses/iou_loss.py:472]: -> None

File: mmdet/models/losses/iou_loss.py [hunk 11]
```diff
@@ -444,18 +466,21 @@ def forward(self,
...
-                target,
-                weight=None,
-                avg_factor=None,
-                reduction_override=None,
+                pred: Tensor,
+                target: Tensor,
+                weight: Optional[Tensor] = None,
+                avg_factor: Optional[int] = None,
+                reduction_override: Optional[str] = None,
                 **kwargs):
```
Comment (BIGWangYuDong) [inline mmdet/models/losses/iou_loss.py:484]: Check whether add  `-> xx`  in all functions

Comment (lihua199710) [reply mmdet/models/losses/iou_loss.py]: OK. I have checked the return value of all functions and modified the docstring.

File: mmdet/models/losses/iou_loss.py [hunk 12]
```diff
@@ -124,7 +136,7 @@ def diou_loss(pred, target, eps=1e-7):
         pred (Tensor): Predicted bboxes of format (x1, y1, x2, y2),
             shape (n, 4).
         target (Tensor): Corresponding gt bboxes, shape (n, 4).
-        eps (float): Eps to avoid log(0).
+        eps (float): Epsilon to avoid log(0).
```
Comment (BIGWangYuDong) [inline mmdet/models/losses/iou_loss.py:139]: should have a blank line before Returns

**Review by BIGWangYuDong — APPROVED** [review]
LGTM

File: mmdet/models/losses/iou_loss.py [hunk 13]
```diff
@@ -443,20 +523,49 @@ def forward(self,
 
 @MODELS.register_module()
 class DIoULoss(nn.Module):
+    r"""`Implementation of Distance-IoU Loss: Faster and Better
```
Comment (ZwwWayne) [inline mmdet/models/losses/iou_loss.py:526]:
```suggestion
    r"""Implementation of `Distance-IoU Loss: Faster and Better
```

File: mmdet/models/losses/iou_loss.py [hunk 14]
```diff
@@ -443,20 +523,49 @@ def forward(self,
...
+    Learning for Bounding Box Regression, https://arxiv.org/abs/1911.08287`_.
```
Comment (ZwwWayne) [inline mmdet/models/losses/iou_loss.py:527]:
```suggestion
    Learning for Bounding Box Regression https://arxiv.org/abs/1911.08287`_.
```

File: mmdet/models/losses/iou_loss.py [hunk 15]
```diff
@@ -49,18 +55,24 @@ def iou_loss(pred, target, linear=False, mode='log', eps=1e-6):
...
     This is an implementation of paper
     `Improving Object Localization with Fitness NMS and Bounded IoU Loss.
     <https://arxiv.org/abs/1711.00164>`_.
 
     Args:
-        pred (torch.Tensor): Predicted bboxes.
-        target (torch.Tensor): Target bboxes.
-        beta (float): beta parameter in smoothl1.
-        eps (float): eps to avoid NaN.
+        pred (Tensor): Predicted bboxes.
```
Comment (ZwwWayne) [inline mmdet/models/losses/iou_loss.py:69]: suggest to also indicate the shape of the tensor

**Review by ZwwWayne — COMMENTED** [review]
For loss functions, it is better to also indicate the shape of the tensor if the module, e.g., IoULoss has strong assumptions about the input.

File: mmdet/models/losses/iou_loss.py (hunk 15 above)
Comment (lihua199710) [reply mmdet/models/losses/iou_loss.py]: Yes, i have specified the shape of inputs.

**Review by ZwwWayne — APPROVED** [review]
